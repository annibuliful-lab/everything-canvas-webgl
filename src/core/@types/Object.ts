export interface I2dPosition {
  x: number;
  y: number;
}

export interface I3dPosition extends I2dPosition {
  y: number;
  w?: number;
}

export interface I2dDimension {
  width: number;
  height: number;
}

export interface I3dDimension extends I2dDimension {
  depth: number;
}

export type ObjectPositioDimension = I2dPosition & I2dDimension;

export interface I2dObject {
  x: number;
  y: number;
  width: number;
  height: number;
  get left(): number;
  get right(): number;
  get top(): number;

  get bottom(): number;
  get topLeft(): I2dPosition;
  get topRight(): I2dPosition;
  get bottomLeft(): I2dPosition;
  get bottomRight(): I2dPosition;
  draw(ctx: CanvasRenderingContext2D): void;
  contains(position: I2dPosition): boolean;
  toObject(): Object;
}
