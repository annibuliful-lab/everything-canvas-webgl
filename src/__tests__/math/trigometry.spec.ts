import { I2dPosition } from "../../core/@types/Object";
import {
  angleToRadian,
  radianToAngle,
  calculateRotatePoint,
  getBoundingBoxByRotate,
} from "../../core/math/trigometry";

describe("trigometry", () => {
  it("it should convert angle to radian", () => {
    expect(angleToRadian(45)).toEqual(0.7853981633974483);
  });

  it("it should convert radian to angle", () => {
    expect(radianToAngle(0.79)).toEqual(45.26366581533504);
  });

  it("it should get new rotated point by 30 degree with origin = (10,10) and point (10,10)", () => {
    expect(
      calculateRotatePoint({
        origin: { x: 10, y: 10 },
        point: { x: 10, y: 10 },
        angle: 30,
      })
    ).toMatchSnapshot();
  });

  it("it should get new rotated point by 30 degree with origin = (1,1) and point (10,10)", () => {
    expect(
      calculateRotatePoint({
        origin: { x: 1, y: 1 },
        point: { x: 10, y: 10 },
        angle: 30,
      })
    ).toMatchSnapshot();
  });

  it("it should get new bounding box by 40 degree with origin = (1,1) and point[(10,10),(30,10),(10,30),(30,30)]", () => {
    const points: [I2dPosition, I2dPosition, I2dPosition, I2dPosition] = [
      { x: 10, y: 10 },
      { x: 30, y: 10 },
      { x: 10, y: 30 },
      { x: 30, y: 30 },
    ];

    expect(
      getBoundingBoxByRotate({ origin: { x: 1, y: 1 }, points, angle: 30 })
    ).toMatchSnapshot();
  });
});
