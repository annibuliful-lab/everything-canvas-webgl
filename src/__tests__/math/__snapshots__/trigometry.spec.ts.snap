// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trigometry it should get new bounding box by 40 degree with origin = (1,1) and point[(10,10),(30,10),(10,30),(30,30)] 1`] = `
Object {
  "bottomLeft": Object {
    "x": -5.70577136594005,
    "y": 40.61473670974872,
  },
  "bottomRight": Object {
    "x": 21.614736709748723,
    "y": 40.61473670974872,
  },
  "topLeft": Object {
    "x": -5.70577136594005,
    "y": 13.294228634059948,
  },
  "topRight": Object {
    "x": 21.614736709748723,
    "y": 13.294228634059948,
  },
}
`;

exports[`trigometry it should get new rotated point by 30 degree with origin = (1,1) and point (10,10) 1`] = `
Object {
  "x": 4.294228634059949,
  "y": 13.294228634059948,
}
`;

exports[`trigometry it should get new rotated point by 30 degree with origin = (10,10) and point (10,10) 1`] = `
Object {
  "x": 10,
  "y": 10,
}
`;
